### PR TITLE
[SPARK-33752][SQL] Avoid the getSimpleMessage of AnalysisException adds semicolon repeatedly

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current `getSimpleMessage` of `AnalysisException` may adds semicolon repeatedly. There show an example below:
`select decode()`

The output will be:
```
org.apache.spark.sql.AnalysisException
Invalid number of arguments for function decode. Expected: 2; Found: 0;; line 1 pos 7
```

### Why are the changes needed?
Fix a bug, because it adds semicolon repeatedly.

### Does this PR introduce _any_ user-facing change?
Yes. the message of AnalysisException will be correct.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by maropu** [comment]
LGTM if the tests pass.

**Comment by beliefer** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 1]
```diff
@@ -48,9 +48,11 @@ class AnalysisException protected[sql] (
 
   // Outputs an exception without the logical plan.
   // For testing only
-  def getSimpleMessage: String = {
+  def getSimpleMessage: String = if (line.isDefined || startPosition.isDefined) {
     val lineAnnotation = line.map(l => s" line $l").getOrElse("")
     val positionAnnotation = startPosition.map(p => s" pos $p").getOrElse("")
     s"$message;$lineAnnotation$positionAnnotation"
+  } else {
+    message
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:56]: You could avoid the branch:
```scala
  val linePosAnnotation = line.map(l => s" line $l").getOrElse("") + startPosition.map(p => s" pos $p").getOrElse("")
  val sep = if (linePosAnnotation.isEmpty) "" else ";"
  s"$message$sep$linePosAnnotation"
```

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala:56]: It's OK too.

**Comment by beliefer** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@beliefer, looks like there's one more test failure left at https://github.com/apache/spark/pull/30724/checks?check_run_id=1554753236

**Comment by beliefer** [comment]
> @beliefer, looks like there's one more test failure left at https://github.com/apache/spark/pull/30724/checks?check_run_id=1554753236

Thank you very much!

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@beliefer, do you mind opening a backport PR for branch-3.1 too? I think it will easily cause some conflicts. I think should at least do it for `branch-3.1` too.

**Comment by beliefer** [comment]
> @beliefer, do you mind opening a backport PR for branch-3.1 too? I think it will easily cause some conflicts. I think should at least do it for `branch-3.1` too.

When will spark 3.1 be released? I'm worried it's too late.

**Comment by beliefer** [comment]
@HyukjinKwon Thanks for your work!
@maropu @MaxGekk Thanks for your review!

**Comment by HyukjinKwon** [comment]
The first release candidate will start from the first week of next month. It will take roughly one month+ historically.

**Comment by beliefer** [comment]
> The first release candidate will start from the first week of next month. It will take roughly one month+ historically.

OK. Let me do it.

**Pull request closed without merging**